File: hw1/HomeWork1/Parser.cs
using System;
using System.Linq;

namespace Homework1
{
    public static class Parser
    {
        private static readonly string[] SupportedOperations = {
            "+",
            "-",
            "*",
            "/"
        };
        
        public static int TryParseArguments(string[] args, out double val1, out string operation, out double val2)
        {
            var isVal1Int = double.TryParse(args[0], out val1);
            operation = args[1];
            var isVal2Int = double.TryParse(args[2], out val2);

            if (!isVal1Int || !isVal2Int)
            {
                Console.WriteLine($"{args[0]}{args[1]}{args[2]} is not a valid calculation syntax");
                return 1;
            }

            if (!SupportedOperations.Contains(operation))
            {
                Console.WriteLine(
                    $"{args[0]}{args[1]}{args[2]} is not a valid calculation syntax. "
                    + $"Supported operations are "
                    + $"{SupportedOperations.Aggregate((c, n) => $"{c} {n}")}");
                return 2;
            }
            
            if (args[1] == "/" && args[2] == "0")
            {
                Console.WriteLine($"{args[0]}{args[1]}{args[2]} - divide by zero");
                return 3;
            }

            return 0;
        }
    }
}

File: hw4/hw4/Parser.fs
namespace hw4

module Parser =

    let isOperationSupported (operation : string) =
        match operation with
            | "+" | "-" | "*" | "/" -> true
            | _ -> false
    let TryParseArguments (args : string[]) (arg1 : outref<int>) (operation : outref<string> ) (arg2 : outref<int>) =
        let isVal1Int = System.Int32.TryParse(args.[0], &arg1)
        let isVal2Int = System.Int32.TryParse(args.[2], &arg2)
        operation <- args.[1]
        if not (isVal1Int && isVal2Int) then
             printfn $"{args.[0]} {args.[1]} {args.[2]} invalid arguments."
             1
        elif not (isOperationSupported args.[1]) then
             printf $"{args.[0]} {args.[1]} {args.[2]} invalid arguments."
             2
        elif args.[1] = "/" && args.[2] = "0" then
             printf $"{args.[0]} {args.[1]} {args.[2]} divide bu zero."
             3
        else 0    

